docs/developer/architecture/configurations.rst

```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
+ - The YAML format utilizes white spaces to define variable hierarchies and as
+   such it is easier to manage with code mean while the flat key-value format
+   requires parsing of flat keys into hierarchical structures which can be
+   tedious.
+
+A sample ConfigMap can be defined with the following fields.
+
+An example of such a ConfigMap:
+
+.. code-block:: yaml
+
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      namespace: <namespace>
+      name: <config-name>
+    data:
+      config.yaml: |-
+        apiVersion: <object-version>
+        kind: <kind>
+        spec:
+          <key>: <values>
+
+**Use case 1:**
+
+Configure and store the number of replicas for service specific Deployments
+found in the `metalk8s-monitoring` namespace using the ConfigMap format.
+
+.. code-block:: yaml
+
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      namespace: metalk8s-monitoring
+      name: metalk8s-grafana-config
+    data:
+      config.yaml: |-
+        apiVersion: metalk8s.scality.com/v1alpha1
+        kind: GrafanaConfig
+        spec:
+          deployment:
+            replicas: 2
+
+Non-goals
+~~~~~~~~~
+
+This section contains requirements stated above which the current design choice
+does not cater for and will be addressed later:
+
+- Persisting newly added Grafana dashboards or new Grafana datasources
+  especially for modifications added via the Grafana UI cannot be stored in
+  ConfigMaps and hence will be catered for later.
+
+- As stated in the requirements, adding and editing Prometheus alert rules
+  is also not covered by the chosen design choice and will need to be addressed
+  differently.
+
+Rejected design choices
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Consul KV vs ConfigMap
+~~~~~~~~~~~~~~~~~~~~~~
+
+This approach offers a full fledge KV store with a /kv endpoint which allows
+CRUD operations on all KV data stored in it.
+Consul KV also allows access to past versions of objects and has an optimistic
+concurrency when manipulating multiple objects.
+
+Note that, Consul KV store was rejected because managing operations such as
+performing full backups, system restores for a full fledged KV system
+requires time and much more efforts than the ConfigMap approach.
+
+Operator (Custom Controller) vs Salt
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Operators are useful in that, they provide self-healing functionalities on a
+reactive basis. When a user changes a given configuration, it is easy to
+reconcile and apply these changes to the in-cluster objects.
+
+The Operator approach was rejected because it is much more complex, requires
+much more effort to realize and there is no real need for applying changes
+using this method because configuration changes are not frequent
+(for a typical MetalK8s admin, changing the number of replicas for a given
+deployment could happen once in 3 months or less) as such, having an operator
+watch for object changes is not significant and not very useful at this point
+in time.
```

Comment:
Good explanation. Maybe add a small paragraph explaining that we will design Salt formulas to be idempotent (as we do for all formulas), making it easy to apply after a config change and "reconcile the desired state", and just loose the "reactive" aspect of it by making this design choice.

Reply:
Config changes in this design and our current approach are not handled by salt I think...it is a Kubernetes thing(kubectl edit). Also, state reconciliation is done by Kubernetes.. no??

Reply:
Sorry if I wasn't clear. In the current approach, we have:

- Salt creates ConfigMaps for all services with default values
- Admin sets the desired configurations by `kubectl edit cm ...` if required (or through CLI/UI)

At that time, nothing is applied to the supposedly configured services.
We need to re-run the Salt formulas (which a CLI/UI could do after editing the ConfigMap).

This implies a two-step process (whereas the Operator approach only requires the CM edition to have changes applied), *unless* we implement a reactive behaviour (through Beacon+Reactor pairs) directly in Salt master(s). _(this makes for a nice benefit of ConfigMaps vs. CRDs, as the former can be mounted as files, and reacting to file changes is rather easy with Salt)_

Reply:
Personnaly prefer to have operator instead of beacon + reactor for this (even if for first iteration having a two-step process with salt is better/simpler IMO)

Reply:
Well, we can easily transform Helm charts into Operators using Operator SDK, could be worth looking into.